com.unity.shadergraph/Editor/Data/Graphs/CategoryData.cs

```diff
@@ -17,37 +17,117 @@ public string name
             get => m_Name;
             set => m_Name = value;
         }
+        public string categoryGuid => this.objectId;
 
         [SerializeField]
-        Guid m_CategoryGuid;
+        List<JsonRef<ShaderInput>> m_ChildObjectList = new List<JsonRef<ShaderInput>>();
 
-        public Guid categoryGuid
+        public RefValueEnumerable<ShaderInput> Children => m_ChildObjectList.SelectValue();
+
+        // We expose the object list as a HashSet of their objectIDs for faster existence checks
+        HashSet<string> m_ChildObjectIDSet = new HashSet<string>();
+
+        public int childCount => m_ChildObjectIDSet.Count;
+
+        public void InsertItemIntoCategory(ShaderInput itemToAdd, int insertionIndex = -1)
         {
-            get => m_CategoryGuid;
-            set => m_CategoryGuid = value;
+            if (itemToAdd == null)
+            {
+                AssertHelpers.Fail("Tried to insert item into category that was null.");
+                return;
+            }
+
+            if (insertionIndex == -1)
+            {
+                m_ChildObjectList.Add(itemToAdd);
+                m_ChildObjectIDSet.Add(itemToAdd.objectId);
+            }
+            else
+            {
+                m_ChildObjectList.Insert(insertionIndex, itemToAdd);
+                // Recreate the hash-set as the data structure does not allow for moving/inserting elements
+                m_ChildObjectIDSet.Clear();
+                foreach (var childObject in m_ChildObjectList)
+                {
+                    m_ChildObjectIDSet.Add(childObject.value.objectId);
+                }
+            }
         }
 
-        // We store Guids as a list out to the graph asset
-        [SerializeField]
-        List<Guid> m_ChildItemIDList;
+        public void RemoveItemFromCategory(ShaderInput itemToRemove)
+        {
+            m_ChildObjectList.Remove(itemToRemove);
+            if (m_ChildObjectIDSet.Contains(itemToRemove.objectId))
+                m_ChildObjectIDSet.Remove(itemToRemove.objectId);
+        }
+
+        public void MoveItemInCategory(ShaderInput itemToMove, int newIndex)
+        {
+            m_ChildObjectList.Remove(itemToMove);
```

Comment:
Any reason this can't just call `RemoveItemFromCategory` then `InsertItemIntoCategory`. There's a lot of error checking you do in Insert in particular that isn't done here (null item, -1 index, etc...)

Reply:
Though I guess one thing that not explicitly calling those existing functions does avoid is having to update the HashSet by adding and removing an item, I think correct-ness trumps performance concerns here

Reply:
You could maybe refactor the validation of the index etc... to a helper, but I think that removing+inserting into a hashset is so trivial it shouldn't matter